src/transformStateWithClones.js

```diff
@@ -7,7 +7,41 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const clonesOfState = [];
+  let stateCopy = { ...state };
+
+  for (let i = 0; i < actions.length; i++) {
+    switch (actions[i].type) {
+      case 'addProperties':
+        Object.assign(stateCopy, actions[i].extraData);
+
+        break;
+
+      case 'removeProperties':
+        for (const key of actions[i].keysToRemove) {
+          delete stateCopy[key];
+        }
+
+        break;
+
+      case 'clear':
+        for (const key in stateCopy) {
+          delete stateCopy[key];
+        }
+
+        break;
+
+      default:
+
+        break;
+    }
+
+    clonesOfState[i] = stateCopy;
+
+    stateCopy = { ...clonesOfState[i] };
```

Comment:
```suggestion
clonesOfState.push({ ...stateCopy }));
```
Don't overcomplicate your code